src/navigation/onboardingStateNavigation.tsx

```diff
@@ -0,0 +1,135 @@
+import Debug from 'debug';
+import React, {ReactNode} from 'react';
+import {State} from 'xstate';
+import {APP_ID} from '../@config/constants';
+import {translate} from '../localization/Localization';
+import {OnboardingContext, OnboardingMachine} from '../machines/onboardingMachine';
+import store from '../store';
+import {ScreenRoutesEnum} from '../types';
+import {
+  IOnboardingMachineContext,
+  IOnboardingPersonalData,
+  OnboardingEvents,
+  OnboardingEventTypes,
+  OnboardingInterpretType,
+  OnboardingStates,
+} from '../types/onboarding';
+import {LOGIN_SUCCESS} from '../types/store/user.action.types';
+import RootNavigation from './rootNavigation';
+
+const debug: Debug.Debugger = Debug(`${APP_ID}:storageService`);
+export const onboardingStateNavigationListener = (
+  onboardingMachine: OnboardingInterpretType,
+  state: State<
+    IOnboardingMachineContext,
+    OnboardingEventTypes,
+    any,
+    {
+      value: any;
+      context: IOnboardingMachineContext;
+    },
+    any
+  >,
+  navigation?: any,
+) => {
+  const onBack = () => onboardingMachine.send(OnboardingEvents.PREVIOUS);
+  const onNext = () => onboardingMachine.send(OnboardingEvents.NEXT);
+  const context = onboardingMachine.getSnapshot().context;
+
+  const nav = navigation ?? RootNavigation;
+  if (nav === undefined || !nav.isReady()) {
+    debug(`navigation not ready yet`);
+    return;
+  }
+  if (state.matches(OnboardingStates.showIntro)) {
+    nav.navigate(ScreenRoutesEnum.WELCOME, {context, onNext});
+  } else if (state.matches(OnboardingStates.acceptAgreement)) {
+    nav.navigate(ScreenRoutesEnum.TERMS_OF_SERVICE, {
+      headerTitle: translate('terms_of_service_title'),
+      context,
+      onBack,
+      onNext,
+      onDecline: () => onboardingMachine.send(OnboardingEvents.DECLINE),
+      onAcceptTerms: (accept: boolean) =>
+        onboardingMachine.send({
+          type: OnboardingEvents.SET_TOC,
+          data: accept,
+        }),
+      onAcceptPrivacy: (accept: boolean) =>
+        onboardingMachine.send({
+          type: OnboardingEvents.SET_POLICY,
+          data: accept,
+        }),
+      isDisabled: () => onboardingMachine.getSnapshot()?.can(OnboardingEvents.NEXT) !== true,
```

Comment:
i am running into this myself and wonder, would it be nicer if we supply buttons to every screen like we already do with some? so then you can have a button prop fully customizable and we can combine some props. Would also make a screen more reusable.

export interface IButton {
  caption: string;
  onPress: () => Promise<void>;
  disabled?: boolean;
}